test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs

```diff
@@ -63,6 +64,164 @@ public void LightGBMBinaryEstimator()
+        }
+
+        [Fact]
+        public void LightGBMMonotonicArgumentPositive()
+        {
+            Dictionary<string, object> options = new Dictionary<string, object>();
+            options["monotone_constraints"] = (positive: new Range[] { new Range(1, 6) }, negative: new Range[] { });
+            var trainer = new LightGbmBinaryTrainer(Env, new Options() { });
+            trainer.ExpandMonotonicConstraints(ref options, 10);
+            Assert.True(options.ContainsKey("monotone_constraints"));
+            var constraintString = options["monotone_constraints"];
+            Assert.Equal("0,1,1,1,1,1,1,0,0,0", constraintString);
+        }
+
+        [Fact]
+        public void LightGBMMonotonicArgumentNegative()
+        {
+            Dictionary<string, object> options = new Dictionary<string, object>();
+            options["monotone_constraints"] = (positive: new Range[] { }, negative: new Range[] { new Range(1, 6) });
+            var trainer = new LightGbmBinaryTrainer(Env, new Options() { });
+            trainer.ExpandMonotonicConstraints(ref options, 10);
+            Assert.True(options.ContainsKey("monotone_constraints"));
+            var constraintString = options["monotone_constraints"];
+            Assert.Equal("0,-1,-1,-1,-1,-1,-1,0,0,0", constraintString);
+        }
+
+        [Fact]
+        public void LightGBMMonotonicArgumentOverlapping()
+        {
+            Dictionary<string, object> options = new Dictionary<string, object>();
+            options["monotone_constraints"] = (positive: new Range[] { new Range(0, 6) }, negative: new Range[] { new Range(2, 4) });
+            var trainer = new LightGbmBinaryTrainer(Env, new Options() { });
+            trainer.ExpandMonotonicConstraints(ref options, 10);
+            Assert.True(options.ContainsKey("monotone_constraints"));
+            var constraintString = options["monotone_constraints"];
+            Assert.Equal("1,1,-1,-1,-1,1,1,0,0,0", constraintString);
+        }
+
+        [Fact]
+        public void LightGBMMonotonicArgumentMultipleRange()
+        {
+            Dictionary<string, object> options = new Dictionary<string, object>();
+            options["monotone_constraints"] = (positive: new Range[] { new Range(0, 4), new Range(6, 8) }, negative: new Range[] { });
+            var trainer = new LightGbmBinaryTrainer(Env, new Options() { });
+            trainer.ExpandMonotonicConstraints(ref options, 10);
+            Assert.True(options.ContainsKey("monotone_constraints"));
+            var constraintString = options["monotone_constraints"];
+            Assert.Equal("1,1,1,1,1,0,1,1,1,0", constraintString);
+        }
+
+
+        [Fact]
+        public void LightGBMMonotonicArgumentExtendToEnd()
+        {
+            Dictionary<string, object> options = new Dictionary<string, object>();
+            options["monotone_constraints"] = (positive: new Range[] { new Range(0, null) }, negative: new Range[] { });
```

Comment:
>negative: new Range[] { } [](start = 93, length = 25)

it looks weird to specify empty range. Is it necessary?

Reply:
I am short circuiting things here for testing. This sets the monotone_constraints parameter in the options dictionary so that I can call ExpandMonotoneConstraints to see what the value is. But as for the user - they dont have to specify both, neither from command line usage or api usage.